internal/deployment_client.go

```diff
@@ -0,0 +1,225 @@
+
+		// When this deployment was created.
+		CreateTime time.Time
+
+		// Whether this deployment is the current one for its deployment series.
+		Current bool
+	}
+
+	// DeploymentListIterator is an iterator for deployments.
+	// NOTE: Experimental
+	DeploymentListIterator interface {
+		// HasNext - Return whether this iterator has next value.
+		HasNext() bool
+
+		// Next - Returns the next deployment and error
+		Next() (*DeploymentListEntry, error)
+	}
+
+	// DeploymentListOptions are the parameters for configuring listing deployments.
+	// NOTE: Experimental
+	DeploymentListOptions struct {
+		// PageSize - How many results to fetch from the Server at a time.
+		// Optional: defaulted to 1000
+		PageSize int
+
+		// SeriesName - Filter with the name of the deployment series.
+		// Optional: If present, use an exact series name match.
+		SeriesName string
+	}
+
+	// DeploymentReachabilityInfo extends DeploymentInfo with reachability information.
+	// NOTE: Experimental
+	DeploymentReachabilityInfo struct {
+		DeploymentInfo
+
+		// Reachability - Kind of tasks that may reach a worker
+		// associated with a deployment.
+		Reachability DeploymentReachability
+
+		// LastUpdateTime - When reachability was last computed. Computing reachability
+		// is an expensive operation, and the server caches results.
+		LastUpdateTime time.Time
+	}
+
+	// DeploymentMetadataUpdate modifies user-defined metadata entries that describe
+	// a deployment.
+	// NOTE: Experimental
+	DeploymentMetadataUpdate struct {
+		// UpsertEntries - Metadata entries inserted or modified. When values are not
+		// of type *commonpb.Payload, the default data converter will be used to generate
+		// payloads.
+		UpsertEntries map[string]interface{}
+
+		// RemoveEntries - List of keys to remove from the metadata.
+		RemoveEntries []string
+	}
+
+	// DeploymentSetCurrentOptions provides options for DeploymentClient.SetCurrent.
+	// NOTE: Experimental
+	DeploymentSetCurrentOptions struct {
+		// Deployment - An identifier for this deployment.
+		Deployment Deployment
+
+		// ClientIdentity - Optional: The identity of the client who initiated this request.
+		ClientIdentity string
+
+		// MetadataUpdate - Optional: Changes to the user-defined metadata entries
+		// for this deployment.
+		MetadataUpdate DeploymentMetadataUpdate
+	}
+
+	// DeploymentSetCurrentResponse is the response type for DeploymentClient.SetCurrent.
+	// NOTE: Experimental
+	DeploymentSetCurrentResponse struct {
+		// Current - Information about the current deployment after this operation.
+		Current DeploymentInfo
+
+		// Previous - Information about the last current deployment, i.e., before this operation.
+		Previous DeploymentInfo
+	}
+
+	// DeploymentClient is the client that manages deployments.
+	// NOTE: Experimental
+	DeploymentClient interface {
+		// Describes an existing deployment.
+		// NOTE: Experimental
+		Describe(ctx context.Context, deploymentID Deployment) (DeploymentInfo, error)
```

Comment:
I wonder if for `Describe` and `GetCurrent` you'd rather have options structures just to be future proof (Go is a bit annoying here w/ no overloads or optional params).

Reply:
Makes sense, it will be a bit more annoying to use but it is worth avoiding a `DescribeWithOptions`...

Reply:
`Deployment` is already a struct, so that will give us a escape hatch if we need a deployment id with other fields. Adding `DeploymentGetCurrentOptions` that was just a string argument...